ddl/syncer.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/terror"
+	goctx "golang.org/x/net/context"
+)
+
+const (
+	ddlAllSchemaVersions   = "/tidb/ddl/all_schem_versions"
+	ddlGlobalSchemaVersion = "/tidb/ddl/global_schem_version"
+	initialVersion         = "0"
+	putKeyDefaultRetryCnt  = 3
+	putKeyNoRetry          = 1
+	putKeyDefaultTimeout   = 3 * time.Second
+	checkVersInterval      = 10 * time.Millisecond
+)
+
+// checkVersFirstWaitTime is used for testing.
+var checkVersFirstWaitTime = 30 * time.Millisecond
+
+type schemaVersionSyncer struct {
+	selfSchemaVerPath string
+	etcdCli           *clientv3.Client
+	LatestVerCh       clientv3.WatchChan
+}
+
+func (s *schemaVersionSyncer) putKV(ctx goctx.Context, retryCnt int, key, val string) error {
+	var err error
+	ctx, cancel := goctx.WithTimeout(ctx, putKeyDefaultTimeout*time.Duration(retryCnt))
+	defer cancel()
+	for i := 0; i < retryCnt; i++ {
+		_, err = s.etcdCli.Put(ctx, key, val)
+		if err == nil {
+			return nil
+		}
+		log.Warnf("put schema version %s failed %v no.%d", val, err, i)
+	}
+	return errors.Trace(err)
+}
+
+func (s *schemaVersionSyncer) Init(ctx goctx.Context) error {
+	_, err := s.etcdCli.Txn(ctx).
+		If(clientv3.Compare(clientv3.CreateRevision(ddlGlobalSchemaVersion), "=", 0)).
+		Then(clientv3.OpPut(ddlGlobalSchemaVersion, initialVersion)).
+		Commit()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	s.LatestVerCh = s.etcdCli.Watch(ctx, ddlGlobalSchemaVersion)
+	return s.putKV(ctx, putKeyDefaultRetryCnt, s.selfSchemaVerPath, initialVersion)
+}
+
+func (s *schemaVersionSyncer) UpdateSelfVersion(ctx goctx.Context, version int64) error {
+	ver := strconv.FormatInt(version, 10)
+	return s.putKV(ctx, putKeyNoRetry, s.selfSchemaVerPath, ver)
```

Comment:
Design document has been a retry, but you later comments said the retry can be removed.
At present I have no particular tendency to which side.